### PR TITLE
Allow users to add custom songs to the music list on a per-area basis

### PR DESCRIPTION
**Pull request by Salanto**
This implements a feature seen in more advanced Tsuserver derivatives, i.E KFO-Server or TsuOLE, allowing CMs to add their songs to the server music list on a per-area basis, which is then sent to clients when joining the area.

The main server list is stored as the "root" list and custom songs are only appended. To avoid conflicts with the alias system root and customs can not have the same songs, but when the root list is disabled any song can be added if it follows the server approved remote content sources or is a local song.

Also comes with tests so you can laugh at me for not considering an edge case in them.

Depends on #223 and #224

### CODE REVIEW

**Comment by Salanto** [comment]
Please squash merge this if possible-

File: core/src/music_manager.cpp [hunk 1]
```diff
@@ -0,0 +1,237 @@
...
+
+bool MusicManager::validateSong(QString f_song_name, QStringList f_approved_cdns)
+{
+    QStringList l_extensions = {".opus", ".ogg", ".mp3", ".wav" };
+
+    bool l_cdn_approved = false;
+    //Check if URL formatted.
+    if (f_song_name.contains("/")) {
+        //Only allow HTTPS/HTTP sources.
+        if (f_song_name.startsWith("https://") | f_song_name.startsWith("http://")) {
```
Comment (in1tiate) [inline core/src/music_manager.cpp:53]: `|` is bitwise OR, please use logical OR (`||`)

File: core/src/music_manager.cpp [hunk 2]
```diff
@@ -0,0 +1,237 @@
...
+            for (const QString &l_cdn : qAsConst(f_approved_cdns)) {
+                //Iterate trough all available CDNs to find an approved match
+                if (f_song_name.startsWith("https://" + l_cdn + "/", Qt::CaseInsensitive)
+                        | f_song_name.startsWith("http://" + l_cdn + "/", Qt::CaseInsensitive)) {
```
Comment (in1tiate) [inline core/src/music_manager.cpp:57]: `|` is bitwise OR, please use logical OR (`||`)

File: core/src/music_manager.cpp [hunk 3]
```diff
@@ -0,0 +1,237 @@
...
+    QString l_real_name = f_real_name;
+    if (f_real_name.split(".").size() == 1) {
+        l_real_name = l_real_name + ".opus";
+    }
+
+    if (!(validateSong(l_song_name, m_cdns) && validateSong(l_real_name, m_cdns))) {
+        return false;
+    }
+
+    //Avoid conflicts by checking if it exits.
```
Comment (in1tiate) [inline core/src/music_manager.cpp:103]:
```suggestion
    // Avoid conflicts by checking if it exists.
```

File: core/src/music_manager.cpp [hunk 4]
```diff
@@ -0,0 +1,237 @@
...
+
+    if (m_custom_lists->value(f_area_id).contains(f_song_name)) {
+        return false;
+    }
+
+    if (m_customs_ordered.value(f_area_id).contains(l_song_name)){
+        return false;
+    }
+
+    //There should be a way to directly insert into the Hash. Too bad.
```
Comment (in1tiate) [inline core/src/music_manager.cpp:116]:
```suggestion
    // There should be a way to directly insert into the QMap. Too bad!
```

File: core/src/music_manager.cpp [hunk 5]
```diff
@@ -0,0 +1,237 @@
...
+    if (f_category_name.split(".").size() > 1) {
+        return false;
+    }
+
+    QString l_category_name = f_category_name;
+    if (!f_category_name.startsWith("==")) {
+        l_category_name = "==" + l_category_name + "==";
+    }
+
+    //Avoid conflicts by checking if it exits.
```
Comment (in1tiate) [inline core/src/music_manager.cpp:136]:
```suggestion
    // Avoid conflicts by checking if it exists.
```

File: core/src/music_manager.cpp [hunk 6]
```diff
@@ -0,0 +1,237 @@
...
+            m_custom_lists->insert(f_area_id,l_custom_list);
+
+            //Updating the list alias too.
+            QStringList l_customs_ordered = m_customs_ordered.value(f_area_id);
+            l_customs_ordered.removeAll(f_songcategory_name);
+            m_customs_ordered.insert(f_area_id, l_customs_ordered);
+
+            emit sendAreaFMPacket(AOPacket("FM",musiclist(f_area_id)), f_area_id);
+            return true;
+        } //Falltrough
```
Comment (in1tiate) [inline core/src/music_manager.cpp:168]:
```suggestion
        } // Fallthrough
```

File: core/src/server.cpp [hunk 7]
```diff
@@ -304,6 +314,15 @@ void Server::broadcast(AOPacket packet, AOPacket other_packet, TARGET_TYPE targe
     }
 }
 
+void Server::unicast(AOPacket f_packet, int f_client_id)
+{
+    AOClient* l_client = getClientByID(f_client_id);
+    if (!(l_client == nullptr)) { // This should never happen, but safety first.
```
Comment (in1tiate) [inline core/src/server.cpp:320]:
```suggestion
    if (l_client != nullptr) { // This should never happen, but safety first.
```
why did you do it like this

File: tests/unittest_music_manager/tst_unittest_music_manager.cpp [hunk 8]
```diff
@@ -0,0 +1,322 @@
...
+     */
+    void sanitiseCustomList();
+
+    /**
+     * @brief Tests the removing of custom songs and categories.
+     */
+    void removeCategorySong();
+
+    /**
+     * @brief Tests the retrival of song information.
```
Comment (in1tiate) [inline tests/unittest_music_manager/tst_unittest_music_manager.cpp:65]:
```suggestion
     * @brief Tests the retrieval of song information.
```

File: tests/unittest_music_manager/tst_unittest_music_manager.cpp [hunk 9]
```diff
@@ -0,0 +1,322 @@
...
+     */
+    void songInformation();
+
+    /**
+     * @brief Tests the retrival of the full musiclist for an area.
```
Comment (in1tiate) [inline tests/unittest_music_manager/tst_unittest_music_manager.cpp:70]:
```suggestion
     * @brief Tests the retrieval of the full musiclist for an area.
```

**Review by in1tiate — CHANGES_REQUESTED** [review]
Writing's not easy. That's why Grammarly can help. This sentence is grammatically correct, but it's wordy, and hard to read. It undermines the writer's message and the word choice is bland. Grammarly's cutting edge technology helps you craft compelling, understandable writing that makes an impact on your reader. Much better. Are you ready to give it a try? Installation is simple and free. Visit Grammarly.com today!

File: core/src/server.cpp (hunk 7 above)
Comment (in1tiate) [reply core/src/server.cpp:320]: actually no wait you could even just do 
```suggestion
    if (!l_client) { // This should never happen, but safety first.
```

File: core/src/config_manager.cpp [hunk 10]
```diff
@@ -25,8 +25,9 @@ QSettings* ConfigManager::m_areas = new QSettings("config/areas.ini", QSettings:
 QSettings* ConfigManager::m_logtext = new QSettings("config/text/logtext.ini", QSettings::IniFormat);
 ConfigManager::CommandSettings* ConfigManager::m_commands = new CommandSettings();
 QElapsedTimer* ConfigManager::m_uptimeTimer = new QElapsedTimer;
-QHash<QString,QPair<QString,float>>* ConfigManager::m_musicList = new QHash<QString,QPair<QString,float>>;
+QMap<QString,QPair<QString,int>>* ConfigManager::m_musicList = new QMap<QString,QPair<QString,int>>;
```
[the remark this answers is not in the document]
Comment (in1tiate) [reply core/src/config_manager.cpp:28]: actually please make this a typedef at least

File: core/src/server.cpp (hunk 7 above)
Comment (Salanto) [reply core/src/server.cpp:320]: Second variation results in a clazy. I'll add your first suggestion one for clarity and a clazy-free life.

**Review by in1tiate — APPROVED** [review]
LOOKS GOOD SORRY
